serve/tests/unittest/test_staged_inference_engine.py

```diff
@@ -0,0 +1,322 @@
+                sampling_params=SamplingParams(temperature=1),
+                stopping_criteria=StoppingCriteria(max_tokens=2),
+            ),
+        ]
+    )
+
+    steps = [engine.step() for _ in range(3)]
+    # execute everything to free server and allow it to destroy
+    for i in range(0,100):
+        engine.step()
+    engine.stop()
+
+    assert len(steps[0].outputs) == 1
+    assert steps[0].outputs[0].request_id == request_id_1
+
+    assert len(steps[1].outputs) == 1
+    assert steps[1].outputs[0].request_id == request_id_1
+
+    assert len(steps[2].outputs) == 2
+    assert (
+        get_output_for_request(steps[2].outputs, request_id_1)
+        .sequences[0]
+        .finish_reason
+        == FinishReason.Length
+    )
+    assert get_output_for_request(steps[2].outputs, request_id_2) is not None
+
+
+def test_multiple_requests_preempt():
+    engine = StagingInferenceEngine(
+        tokenizer_module=DummaryTokenizerModule(),
+        model_module_loader=DummaryModelModule,
+        model_module_loader_kwargs = {
+            "max_cached_tokens": 30
+        }
+        )
+    engine.start()
+
+    request_id_1 = "1"
+    request_id_2 = "2"
+
+    engine.add(
+        [
+            Request(
+                request_id=request_id_1,
+                messages=create_messages("test " * 10),
+                sampling_params=SamplingParams(temperature=1),
+                stopping_criteria=StoppingCriteria(max_tokens=7),
+            ),
+        ]
+    )
+
+    engine.add(
+        [
+            Request(
+                request_id=request_id_2,
+                messages=create_messages("test " * 10),
+                sampling_params=SamplingParams(temperature=1),
+                stopping_criteria=StoppingCriteria(max_tokens=7),
+            ),
+        ]
+    )
+
+    steps = [engine.step() for _ in range(20)]
+    # execute everything to free server and allow it to destroy
+    for i in range(0,100):
+        engine.step()
+    engine.stop()
+
+    # Due to asynchronious nature of request submission and processing, we cannot
+    # track exactly on each step certain request is processed
+    # but we can catch a pattern that it is two requests processed, then 1 and then again 2
+    stage = 0
+
+    for s in steps:
+        if stage == 0 and len(s.outputs) == 2:
+            stage = 1
+        if stage == 1 and len(s.outputs) == 1:
+            stage = 2
+        if stage == 2 and len(s.outputs) == 2:
+            stage = 3
+        if stage == 3 and len(s.outputs) == 1:
+            stage = 4
+            assert s.outputs[0].is_finished
+        if stage == 4 and len(s.outputs) > 1:
+            stage = 5
+
+    assert stage == 4
+
+
+# Test to verify if evicted request from active batch which in intermediate
+# state exceeding the max_num_batched_tokens can be processed successfully and will
+# not hang the server in infinite attempt to return it back to the active loop
+def test_cache_evict_hang_staging():
```

Comment:
I confirmed that this test hangs without the engine changes in this PR, and I get the hang warning you added in https://github.com/octoml/mlc-llm/pull/91 as expected. Great! cc @jroesch @itay